File: android/src/main/kotlin/com/uax/liquidflutter/LiquidExtension.kt
package com.uax.liquidflutter

import asia.liquidinc.ekyc.applicant.external.result.LiquidChipIdentificationResult
import asia.liquidinc.ekyc.applicant.external.result.LiquidProcessingResult
import asia.liquidinc.ekyc.applicant.external.result.LiquidProcessingResultStatus
import asia.liquidinc.ekyc.applicant.external.result.chip.ChipIdentificationResultStatus
import asia.liquidinc.ekyc.applicant.external.result.chip.Sex

fun Sex.stringValue(): String {
    return when (this) {
        Sex.MALE -> "0"
        Sex.FEMALE -> "1"
        Sex.OTHERS -> "2"
        else -> "unknown"
    }
}

fun LiquidProcessingResult.toJson(): Any {
    return hashMapOf(
        "status" to result.stringValue(),
        "resultCode" to errorCode,
        "additionalData" to hashMapOf(
            "title" to additionalDataTitle,
            "message" to additionalDataMessage
        )
    )
}

fun LiquidProcessingResultStatus.stringValue(): String {
    return when (this) {
        LiquidProcessingResultStatus.USER_CANCEL -> "userCancel"
        LiquidProcessingResultStatus.MAINTENANCE -> "maintenance"
        LiquidProcessingResultStatus.ERROR -> "error"
        LiquidProcessingResultStatus.SESSION_TIMEOUT -> "sessionTimeout"
        LiquidProcessingResultStatus.TERMS_DO_NOT_AGREE -> "termsDoNotAgree"
        LiquidProcessingResultStatus.COMMUNICATION_FAILURE -> "communicationFailure"
        LiquidProcessingResultStatus.SUCCESS -> "success"
        else -> "unknown"
    }
}

fun LiquidChipIdentificationResult.toJson(): Any {
    return hashMapOf(
        "status" to resultStatus.stringValue(),
        "resultCode" to errorCode,
        "additionalData" to hashMapOf(
            "title" to additionalDataTitle,
            "message" to additionalDataMessage
        )
    )
}

fun ChipIdentificationResultStatus.stringValue(): String {
    return when (this) {
        ChipIdentificationResultStatus.USER_CANCEL -> "userCancel"
        ChipIdentificationResultStatus.MAINTENANCE -> "maintenance"
        ChipIdentificationResultStatus.ERROR -> "error"
        ChipIdentificationResultStatus.SESSION_TIMEOUT -> "sessionTimeout"
        ChipIdentificationResultStatus.SCREEN_TIMEOUT -> "screenTimeout"
        ChipIdentificationResultStatus.SUCCESS -> "success"
        ChipIdentificationResultStatus.PERMISSION_NOT_ALLOWED -> "permissionNotAllowed"
        ChipIdentificationResultStatus.UNSUPPORTED_DEVICE_NFC -> "unsupportedChip"
        ChipIdentificationResultStatus.IC_CHIP_FORGOT_PIN -> "chipForgotPin"
        ChipIdentificationResultStatus.IC_CHIP_LOCKED -> "chipLocked"
        ChipIdentificationResultStatus.IC_CHIP_EXPIRED -> "chipExpired"
        ChipIdentificationResultStatus.IC_CHIP_IDENTIFY_DENIED -> "chipIdentifyDenied"
        ChipIdentificationResultStatus.IC_CHIP_IDENTIFY_ERROR -> "chipIdentifyError"
        else -> "unknown"
    }
}

File: ios/Classes/LiquidExtension.swift
import Foundation
import Liquid

extension ProcResult {
    func toJson() -> SdkData {
        return [
            "status": status.stringValue,
            "resultCode": resultCode,
            "additionalData" : [
                "title": additionalData?.maintenanceTitle,
                "message": additionalData?.maintenanceMessage,
            ],
        ]
    }
}

extension Liquid.Sex {
    var stringValue: String {
        switch self {
        case .male:
            return "0"
        case .female:
            return "1"
        case .others:
            return  "2"
        @unknown default:
            return "unknown"
        }
    }
}

extension Liquid.ResultStatus {
    var stringValue: String {
        switch self {
        case .success:
            return "success"
        case .maintenance:
            return "maintenance"
        case .permissionNotAllowed:
            return "permissionNotAllowed"
        case .screenTimeout:
            return "screenTimeout"
        case .sessionTimeout:
            return "sessionTimeout"
        case .ocrInProgress:
            return "ocrInProgress"
        case .ocrUnSupported:
            return "ocrUnSupported"
        case .userCancel:
            return "userCancel"
        case .error:
            return "error"
        case .chipLocked:
            return "chipLocked"
        case .chipUnusual:
            return "chipUnusual"
        case .chipUnusualUpdated:
            return "chipUnusualUpdated"
        case .chipVerifyFailure:
            return "chipVerifyFailure"
        case .chipUnusualResidencecard:
            return "chipUnusualResidencecard"
        case .chipMissingExternalChar:
            return "chipMissingExternalChar"
        case .unsupportedChip:
            return "unsupportedChip"
        case .chipPinFailure:
            return "chipPinFailure"
        case .chipExpired:
            return "chipExpired"
        case .chipForgotPin:
            return "chipForgotPin"
        case .chipIdentifyDenied:
            return "chipIdentifyDenied"
        case .chipIdentifyError:
            return "chipIdentifyError"
        case .termsDoNotAgree:
            return "termsDoNotAgree"
        case .communicationFailure:
            return "communicationFailure"
        @unknown default:
            return "unknown"
        }
    }
}
